### PR TITLE
Remove RELOCATED index shard state

### PR DESCRIPTION
**Pull request by ywelsch**
Relates https://github.com/elastic/elasticsearch/pull/28004#issuecomment-354814578

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/shard/IndexShard.java [hunk 1]
```diff
@@ -581,9 +573,8 @@ public void relocated(
                     consumer.accept(primaryContext);
                     synchronized (mutex) {
                         verifyRelocatingState();
-                        changeState(IndexShardState.RELOCATED, reason);
+                        replicationTracker.completeRelocationHandoff(); // make changes to primaryMode flag only under mutex
```
Comment (bleskes) [inline server/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: question - why is this moved under mutex?

File: server/src/main/java/org/elasticsearch/index/shard/IndexShard.java [hunk 2]
```diff
@@ -1426,20 +1414,16 @@ public boolean isReadAllowed() {
...
-                throw new IllegalIndexShardStateException(shardId, state, "operation only allowed when shard state is one of " + writeAllowedStatesForPrimary + ", origin [" + origin + "]");
-            }
-        } else if (origin.isRecovery()) {
+        if (origin.isRecovery()) {
             if (state != IndexShardState.RECOVERING) {
                 throw new IllegalIndexShardStateException(shardId, state, "operation only allowed when recovering, origin [" + origin + "]");
             }
         } else {
-            assert origin == Engine.Operation.Origin.REPLICA;
-            verifyReplicationTarget();
```
Comment (bleskes) [inline server/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: Isn't `verifyReplicationTarget` still good to check?

**Review by bleskes — COMMENTED** [review]
This is almost too easy :) . I left some questions.

File: server/src/main/java/org/elasticsearch/index/shard/IndexShard.java (hunk 1 above)
Comment (ywelsch) [reply server/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: The primary mode in the replicationTracker is dependent on the shard state and updated together with the shard state (for example in updateShardState). I think that it's good if this state is always modified atomically with the shard state, to ensure that these two never go out of sync. I'm not aware of a specific case where the two could go badly out of sync, so this is more of a pre-cautionary measure. Besides `updateShardState`, where we already update them under the mutex, the two places where primary mode is also updated is when activating a primary on relocation and when completing a relocation (this place here). Both are rarely called and terminate very quickly, so I felt it safer to move them under the mutex.

File: server/src/main/java/org/elasticsearch/index/shard/IndexShard.java (hunk 2 above)
Comment (ywelsch) [reply server/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: good catch. I missed this.

File: server/src/main/java/org/elasticsearch/index/shard/IndexShard.java (hunk 1 above)
Comment (bleskes) [reply server/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: +1

**Review by bleskes — APPROVED** [review]
LGTM

File: server/src/main/java/org/elasticsearch/index/shard/IndexShard.java (hunk 1 above)
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: ++

**Review by dnhatn — COMMENTED** [review]
Late to the party. LGTM!
